mozilla_django_oidc/views.py

```diff
@@ -63,7 +63,16 @@ def get(self, request):
             # Make sure that nonce is not used twice
             del request.session['oidc_nonce']
 
-        if 'code' in request.GET and 'state' in request.GET:
+        if request.GET.get('error'):
+            # Ouch! Something important failed.
+            # Make sure the user doesn't get to continue to be logged in
+            # otherwise the refresh middleware will force the user to
+            # redirect to authorize again if the session refresh has
+            # expired.
+            if is_authenticated(request.user):
+                auth.logout(request)
+            assert not is_authenticated(request.user)
```

Comment:
I think this sounds ok. In https://github.com/mozilla/mozilla-django-oidc/issues/153#issuecomment-317597945 I suggested we add a failure url that's exempt from the middleware and redirect to that.

I don't know offhand if the error message that comes from the SSO is always user-friendly or not. I wonder if it's always the case that an error from the SSO indicates that the user should get logged out of the Django site? Are there any other possible situations?

Reply:
Yeah, this is tricky. First of all, I can't talk about if the SSO ever redirects to `yoursite/oidc/callback/?error=something` where the error isn't "major". 

An alternative solution is to do something like this in the RefreshIDToken middleware:

```python
if request.path == self.failure_url() and request.GET.get('error'):
    # Don't redirect to refresh the ID token! Something's up!
else:
   # the existing regular stuff
```

The counter-argument is that if Auth0 does redirect to the callback with an error it sounds pretty severe and correctly the user should perhaps be kicked out. In my solution, what would happen from a user perspective is "What? I was logged in last Friday and now I no longer am. Oh well, I'll just log in again" and when you do that if Auth0 has an issues they'll be dealt with there. In the case of the Airmozilla failure it seemed the error was that [I hadn't 2FA'ed in a while](https://github.com/mozilla/airmozilla/issues/827#issuecomment-374274903).

Currently, if you try go to `https://www.standu.ps/oidc/callback/?error=something&error_message=Blablabla` you're just redirected back to `https://www.standu.ps/` and that `error` ad `error_message` strings are lost. [Any interesting query string is discarded.](https://github.com/mozilla/mozilla-django-oidc/blob/da3d49a275c1f7bfba49a1f6b9d017090a566325/mozilla_django_oidc/views.py#L83)

Reply:
Also, suppose we had a way of making a dedicated "OIDC callback error page" where RPs can localize the content and potentially display (escaped!) the `error` and `error_message` strings then what will the user do with it? Read it? 
The user just wants to log in and if he was already logged in, he'd have to click the sign in button again. Or sign out. And if the sign in page (e.g. `https://air.mozilla.org/authentication/signin/`) is outside that exempt scheme mentioned above the redirect would start again because of the RefreshIDToken middleware.

Reply:
Would really appreciate some feedback here @johngian @akatsoulas

Reply:
I think the simplest solution would be to just logout the user. Displaying the error message probably won't be that useful to the user. Moreover, for me it makes sense to logout the user if there is a failure in the refresh process

Reply:
@akatsoulas Is that an approval? :)

Reply:
I didn't do a review because @willkg did all the hard work here. Logging the user out sounds good to me